# **YEAST TRANSPOSON CANONICAL SEQUENCES**

Here we present an open access library of transposable element canonical sequences for species in the genus Saccharomyces (Ty elements). Currently, this library contains sequences of full-length elements for major LTR retrotransposon families and subfamilies in S. cerevisiae and S. paradoxus.

The library here is derived from a related resource reported in in Carr et al. (2012), later adapted for re-use by Nelson, Linheiro, and Bergman (2017) and Czaja et al (2020). We decided to update the old Ty query library for the following limitations: (i) the old Ty sequences were predominately adapted from the RepeatMasker Repbase library, which is currently closed access; (ii) the ultimate provenance of some sequences was unclear (i.e. Ty1, Ty2) and some Ty sequences were adopted from legacy Genbank records which may have poor sequence quality (i.e. Ty3, Ty4); (iii) "full-length" Ty sequences in the old library were formatted with LTR and internal regions split from one another (to be used as input for RepeatMasker), rather than true full-length elements; (v) the species of origin for non-cerevisiae Ty elements was inconsistently labeled (i.e. Ty3_1p, Ty5); (vi) the old query sequences lacked systematic structural annotations (coordinates of LTRs and ORFs); and (vi) representatives of described subfamilies (i.e. Ty1') and newly-discovered Ty families (i.e. Tsu4) were not included.

In order to overcome these limitations, we created a comprehensive library of full-length canonical Ty elements for S. cerevisiae and S. paradoxus derived from high-quality public long-read assemblies. To select Ty elements for this library, we first applied a RepeatMasker-based Ty annotation pipeline (described in Czaja et al. 2020) to high quality public genome assemblies for S. cerevisiae and S. paradoxus (Czaja et al. 2020; Yue et al. 2017; Eberlein et al. 2019; Bendixsen et al. 2021; Zhang and Emerson 2019). We next ran LTRharvest and LTRdigest (version and parameters, Ellinghaus, Kurtz, and Willhoeft 2008) on the same set of yeast genome assemblies to generate de novo predictions full-length LTR elements were applied. We then generated multiple sequence alignments for each Ty family (Ty1, Ty2, Ty3, Ty3_1p, Ty4, Tsu4 and Ty5) using all full-length elements detected by the RepeatMasker based pipeline plus the query element used in the Czaja Ty libraries. Multiple alignments were then used to generate neighbor-joining phylogenetic trees with Seaview (version and parameters, Gouy, Guindon, and Gascuel 2010). We then selected a representative full-length genomic replacement that was as close to the original query element in the tree as possible and was supported by a LTRharvest prediction. We also included genomic representatives for several Ty families (e.g Ty5c, Ty1par) and subfamilies (Ty1’, canonical Ty1) with LTRharvest support that were not previously included in the Czaja et al. Ty libraries. Finally, we integrated structural annotations, including flanking LTRs and ORFs (GAG and POL) for all genomic replacements. The coordinates of 5’ and 3’ LTR were adopted from LTRharvest results, which were then validated with RepeatMasker annotations to make sure the length of annotated LTRs were consistent in two independent tools. The ORF prediction was performed with NCBI ORFfinder (Linux x64 version 0.4.3; Minimum length of ORF was set to 600; Allowed start codons were set to “ATG only” for GAG prediction and  “any sense codon” for POL prediction; https://www.ncbi.nlm.nih.gov/orffinder/). The final updated file of Ty queries can be found in File X and annotations in GFF format can be found in File Y. 

<p align="center">
    <img src="https://github.com/bergmanlab/yeast-transposons/blob/main/img/master_lib.jpg?raw=true" alt="Ty library"/>
</p>
